File: java/findModeInBinarySearchTree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    //keeps track of the number of repetitions in the mode so far
    int maxCount = 0;
    //keeps track of the number of repetitions in the values we're currently examining
    int curCount = 0;
    //keeps track of the previous value we looked at in order to compare
    int prevVal;
    //stores the modes and clears them when maxCount increases
    List<Integer> modes = new ArrayList();
    
    public int[] findMode(TreeNode root) {
        //traverse the tree in order to collect the modes
        visit(root);
        //convert modes into an array
        int[] result = new int[modes.size()];
        for (int i = 0; i < modes.size(); i++) {
            result[i] = modes.get(i);
        }
        return result;
    }
    
    private void visit(TreeNode node) {
        //recursive base case
        if (node == null) {
            return;
        }
        //visit left subtree -> root -> right subtree for in-order traversal
        visit(node.left);
        //increment current count if appropriate
        if (node.val == prevVal) {
            curCount++;
        } else {
            curCount = 1;
        }
        //reassign previous value
        prevVal = node.val;
        //if the current value is a mode, add it to the mdoes list
        if (curCount == maxCount) {
            modes.add(node.val);
        //if the current value is greater than the previous modes, clear the modes list then add it
        } else if (curCount > maxCount) {
            maxCount = curCount;
            modes.clear();
            modes.add(node.val);
        }
        visit(node.right);
    }
}

File: python/find_mode_in_binary_search_tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def findMode(self, root: TreeNode) -> List[int]:
        #modes list will be cleared everytime a higher mode is found
        self.modes = []
        #keep track of previous node in order to compare
        self.prevNode = None
        #highest number of repetitions found so far
        self.maxCount = 0
        #current number of repetitions
        self.curCount = 0
        #traverse the tree in-order to count each element in order
        self.visit(root)
        return self.modes
    
    def visit(self, node: TreeNode) -> None:
        if node is None:
            return
        #order: left subtree -> root -> right subtree
        self.visit(node.left)
        #if the current node has the same value as the previous one, update the current count
        if self.prevNode is not None and node.val == self.prevNode.val:
            self.curCount += 1
        else: 
            self.curCount = 1
        #if the current count equals the max count, add the current node's value to the results
        if self.curCount == self.maxCount:
            self.modes.append(node.val)
        #if the current count exceeds the max count, clear the results and add the current node's value
        elif self.curCount > self.maxCount:
            self.maxCount = self.curCount
            self.modes = [node.val]
        self.prevNode = node
        self.visit(node.right)